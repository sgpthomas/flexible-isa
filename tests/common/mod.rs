use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
