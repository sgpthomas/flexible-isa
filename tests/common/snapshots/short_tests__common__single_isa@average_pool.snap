---
source: tests/common/mod.rs
description: generate isa for average_pool
---
[
    (
        2,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?x4) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> ?x5))) (cast<int32> 0))))))) (cast<int32> ?x6))) (cast<int32> (+ (cast<int32> ?x7) (cast<int32> (+ (cast<int32> ?x8) (cast<int32> ?x9))))))",
    ),
    (
        8,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))))) (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> ?#8) (cast<int32> (* (cast<int32> ?#9) (cast<int32> ?x3))))) (cast<int32> (* (cast<int32> ?x4) (cast<int32> ?x5))))) (cast<int32> (* (cast<int32> ?x6) (cast<int32> ?x7))))) (cast<int32> ?#10))))",
    ),
    (
        28,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> (* (cast<int32> ?#1) (cast<int32> (+ (cast<int32> ?#2) (cast<int32> ?#3))))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?#4) (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?#7) (cast<int32> (+ (cast<int32> ?#8) (cast<int32> ?#9))))) (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> ?#10) (cast<int32> (* (cast<int32> ?#11) (cast<int32> ?#12))))) (cast<int32> (* (cast<int32> ?#13) (cast<int32> ?#14))))) (cast<int32> (* (cast<int32> ?#15) (cast<int32> ?#16))))) (cast<int32> ?#17))))))))))",
    ),
    (
        34,
        1,
        "(+ ?x0 ?x1)",
    ),
    (
        16,
        2,
        "(+ (cast<int32> (call[min] (cast<int32> (* (cast<int32> ?#2) (cast<int32> 128))) (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 128))))))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))))) (cast<int32> ?x1))))",
    ),
    (
        17,
        2,
        "(+ (cast<int32> (call[min] (cast<int32> (* (cast<int32> ?#2) (cast<int32> 256))) (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 256))))))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))))) (cast<int32> ?x1))))",
    ),
    (
        18,
        2,
        "(+ (cast<int32> (call[min] (cast<int32> (* (cast<int32> ?#2) (cast<int32> 512))) (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 512))))))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))))) (cast<int32> ?x1))))",
    ),
    (
        23,
        3,
        "(+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))) (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> ?#9) (cast<int32> (* (cast<int32> ?#10) (cast<int32> ?x4))))) (cast<int32> (* (cast<int32> ?x5) (cast<int32> ?x6))))) (cast<int32> (* (cast<int32> ?x7) (cast<int32> ?x8))))) (cast<int32> ?#11))))))",
    ),
    (
        36,
        3,
        "(- (cast<int32> (- (cast<int32> (- (cast<int32> (- (cast<int32> ?#5) (cast<int32> (* (cast<int32> ?#6) (cast<int32> ?x0))))) (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))))) (cast<int32> (* (cast<int32> ?x3) (cast<int32> ?x4))))) (cast<int32> ?#7))",
    ),
    (
        55,
        3,
        "(<= (cast<int32> ?x0) (cast<int32> ?#1))",
    ),
    (
        57,
        3,
        "(get ?x0 (cast<int32> ?x1))",
    ),
    (
        58,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        59,
        3,
        "(get ?x0 (cast<int32x256> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 256))))",
    ),
    (
        60,
        3,
        "(get ?x0 (cast<int32x512> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 512))))",
    ),
    (
        48,
        7,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?x4) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> ?x5))) (cast<int32> 0))))))) (cast<int32> ?x6))",
    ),
    (
        52,
        7,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        41,
        8,
        "(- (cast<int32> (call[max] (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> ?x4))) (cast<int32> (call[min] (cast<int32> (- (cast<int32> (+ (cast<int32> ?x5) (cast<int32> ?x0))) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> ?x4))))) (cast<int32> 0))) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> ?x4))) (cast<int32> 0))))",
    ),
    (
        32,
        9,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        50,
        15,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        42,
        16,
        "(- (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))) (cast<int32> (+ (cast<int32> ?x4) (cast<int32> ?x5))))) (cast<int32> (call[max] (cast<int32> ?x5) (cast<int32> (* (cast<int32> ?x1) (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))))))))",
    ),
]
